### PR TITLE
Variable sensitivity domain

### PR DESCRIPTION
**Pull request by martin-cs**
This PR merges in several years of work on a new domain for the abstract interpreter.  It should solve a number of the long-standing problems with soundness and precision.  The key feature of the domain is that the abstraction used for each type can be selected at run-time, allowing a wide range of scalability / precision trade-offs and easy specialisation without having to re-write all of the array, field, pointer sensitivity, etc.

1. This is the work of a lot of people, notably @thk123 @chrisr-diffblue @hannes-steffenhagen-diffblue @danpoe @NlightNFotis and others.  I have tried to preserve authorship where-ever possible.

2. I don't think it is reasonable to try to review this in the normal way, nor do I think it is viable to split it up further.  Whether it gets merged or not has to be a policy question really.  Over the years this has been going on I have asked @peterschrammel and @tautschnig about this as an approach and they seemed supportive because...

3. The PR should have very little impact on existing functionality, it is enabled with an option in goto-analyzer.  It has also been deployed in at least two industrial applications.

4. There is still work to be done in tidying up the code but I think it can be and would be best done as PRs to it after merging.  I will try to do some during the discussion / review.

5. This includes https://github.com/diffblue/cbmc/pull/5202 and https://github.com/diffblue/cbmc/pull/5204 which were previous attempts at splitting this out a bit.

6. I don't know who is currently looking after
https://github.com/diffblue/cbmc/tree/variable-sensitivity-domain
and
https://github.com/diffblue/cbmc/tree/variable-sensitivity-with-get-function-pointers
( @chrisr-diffblue ? @danpoe ? @hannes-steffenhagen-diffblue ? )
but this PR includes the majority of the content of these.  I have a branch of the remainder and am happy to work with people to rebase that on top of this PR.

Let me know what you think!

### CODE REVIEW

**Comment by martin-cs** [comment]
The unit tests currently don't compile (oops!).  I note that "AWS CodeBuild us-east-1 (cbmc-windows-cmake)" passes; is it supposed to skip / ignore the unit tests?

**Comment by peterschrammel** [comment]
Happy to review if you make the monster pass CI first.

**Comment by peterschrammel** [comment]
> is it supposed to skip / ignore the unit tests?

It probably isn't. @hannes-steffenhagen-diffblue ?

**Comment by martin-cs** [comment]
@peterschrammel Thank you, I think it is just the unit tests which are outstanding.  All the regression tests and linting, formatting, etc. all pass.  I will update the usage of `exprt` in the unit tests and ping you.

**Comment by hannes-steffenhagen-diffblue** [comment]
@martin-cs @peterschrammel The cmake windows build is currently not running any tests actually. This is a long outstanding bug which is one of the things we're hoping to "fix" with our "move CI to github actions" work.

**Comment by martin-cs** [comment]
@hannes-steffenhagen-diffblue cool, as long as this isn't news to you!

**Comment by martin-cs** [comment]
@hannes-steffenhagen-diffblue you probably already know this but it looks like "AWS CodeBuild us-east-1 (cbmc-clang-3-8)" doesn't run any tests either.

**Comment by martin-cs** [comment]
@peterschrammel All of CI now passes, any kind of review would be most appreciated.

File: regression/goto-analyzer-simplify/simplify-lhs-pointer-array-dereference/main.c
Comment (peterschrammel) [inline regression/goto-analyzer-simplify/simplify-lhs-pointer-array-dereference/main.c]: :pick: I don't think this reference is publically accessible.

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_enviroment.cpp]: ❗ Outdated comment headers. This needs replacing - if not in this PR than immediately as a follow-up.

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_enviroment.cpp]: :pick:
```suggestion
 As abstractions may split the assignment into multiple writes (for
```

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_enviroment.cpp:191]: ❓ What is an `i-value`?

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_enviroment.cpp]:
```suggestion
   * 1. split up conjuncts, handle each part separately
```

File: src/analyses/variable-sensitivity/abstract_enviroment.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_enviroment.h:115]: When converting the docs, please move interface docs into the header files. This would make it much easy to understand what a class is doing.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_object.h:3]: Please convert to doxygen.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_object.h]:
```suggestion
 bottom).  Within the hierarchy of objects under it, child classes are
```

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_object.h:259]: Use `///` stile of doxygen comments.

File: src/analyses/variable-sensitivity/array_abstract_object.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/array_abstract_object.cpp]: ❓ The value doesn't seem to be `top` when you pass in `bottom`?

File: src/analyses/variable-sensitivity/constant_array_abstract_object.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/constant_array_abstract_object.cpp:313]: ❓ Not sure about this comment. Is there a test that exposes a related problem? Or it should at least be explained why the current code is wrong.

File: src/analyses/variable-sensitivity/constant_array_abstract_object.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/constant_array_abstract_object.cpp]: ⛏️ Preferred for consistency with the rest of the codebase:
```suggestion
  for(const auto &object : view)
```

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/abstract_object.h:47]: That's quite a generic macro name for something relatively specific.

File: src/analyses/variable-sensitivity/data_dependency_context.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/data_dependency_context.cpp]:
```suggestion
 * additional actions after the base abstract_object_merge has completed its
```

File: src/analyses/variable-sensitivity/full_struct_abstract_object.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/full_struct_abstract_object.cpp]: :pick: put increment into for?

File: src/analyses/variable-sensitivity/interval_abstract_value.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/interval_abstract_value.cpp:214]: 💡  Consider breaking down this giant function.

File: src/analyses/variable-sensitivity/interval_array_abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/interval_array_abstract_object.h:29]: This macro looks really scary ;)

File: src/analyses/variable-sensitivity/struct_abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/struct_abstract_object.h:83]: ❓ 
```suggestion
    const abstract_object_pointert &value,
```

File: src/analyses/variable-sensitivity/struct_abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/struct_abstract_object.h:133]: ❓

File: src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp:12]: No need to repeat this here.

File: src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp]:
```suggestion
    // its arguments as it goes.  A better solution would be to refactor
```

File: src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp]: Please document why this is commented out, or remove.

File: src/analyses/variable-sensitivity/value_set_abstract_object.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/value_set_abstract_object.h:128]: 👍  Here the format is right.

File: src/analyses/variable-sensitivity/variable_sensitivity_domain.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/variable_sensitivity_domain.h]: Concretely refer to the module/class of the existing solution.

File: src/analyses/variable-sensitivity/variable_sensitivity_object_factory.h
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/variable_sensitivity_object_factory.h:77]: ❓ Do we know why it doesn't work?

File: src/analyses/variable-sensitivity/write_location_context.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/write_location_context.cpp]:
```suggestion
 * propagating the update to any children of this abstract object.
```

File: src/analyses/variable-sensitivity/write_location_context.cpp
Comment (peterschrammel) [inline src/analyses/variable-sensitivity/write_location_context.cpp]:
```suggestion
 * \return a clone of this abstract object with its location context
```

File: src/goto-analyzer/goto_analyzer_parse_options.h
Comment (peterschrammel) [inline src/goto-analyzer/goto_analyzer_parse_options.h:133]: 💡 `dependency-graph` seems to be more common.

File: unit/analyses/variable-sensitivity/abstract_object/merge.cpp
Comment (peterschrammel) [inline unit/analyses/variable-sensitivity/abstract_object/merge.cpp]: ❗ 
```suggestion
 Author: Diffblue Limited. 
```

File: unit/analyses/variable-sensitivity/constant_abstract_value/merge.cpp
Comment (peterschrammel) [inline unit/analyses/variable-sensitivity/constant_abstract_value/merge.cpp]: ❗

File: unit/analyses/variable-sensitivity/constant_array_abstract_object/merge.cpp
Comment (peterschrammel) [inline unit/analyses/variable-sensitivity/constant_array_abstract_object/merge.cpp]: ❗ (and all other files)

File: unit/analyses/variable-sensitivity/eval.cpp
Comment (peterschrammel) [inline unit/analyses/variable-sensitivity/eval.cpp]: ❗

**Review by peterschrammel — COMMENTED** [review]
A shallow review of 20KLOC to scan for red flags. The PR is in pretty good state code quality-wise - ie acceptable to be merged. The main concern are outdated documentation formats that need to be fixed so that the doxygen documentation is produced correctly. Also, some copyright headers don't follow the guidelines and must be fixed before merge.

File: regression/goto-analyzer-simplify/simplify-lhs-pointer-array-dereference/main.c
Comment (martin-cs) [reply regression/goto-analyzer-simplify/simplify-lhs-pointer-array-dereference/main.c]: Removed.

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_enviroment.cpp]: I have reformatted all of them.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_object.h:3]: Done.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_object.h:259]: I thought this was an acceptable style?  I thought it was used elsewhere in the code?

File: src/goto-analyzer/goto_analyzer_parse_options.h
Comment (martin-cs) [reply src/goto-analyzer/goto_analyzer_parse_options.h:133]: Yes but this is what the class in question is called and it has been a command-line option for some time.  I'm happy to change it but I don't think it should be part of this PR.

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_enviroment.cpp]: Fixed.  Thanks.

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_enviroment.cpp:191]: ? Is this a font issue? The source says `l-value` as in https://en.wikipedia.org/wiki/Value_(computer_science)#lrvalue

File: src/analyses/variable-sensitivity/abstract_enviroment.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_enviroment.cpp]: Fixed.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_object.h]: Fixed.

File: src/analyses/variable-sensitivity/array_abstract_object.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/array_abstract_object.cpp]: The code is wrong but the way it is used will (I think) mean this case should never come up.  There is work to do on write stacks and writing to an index so I will add this to the TODO for that if it's OK with you?

File: src/analyses/variable-sensitivity/constant_array_abstract_object.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/constant_array_abstract_object.cpp:313]: It's not that the current code is wrong, more that it could be improved.  If your index is not resolvable to a constant then there are a number of possible approaches.  Again, it is on my "to improve" list.

File: src/analyses/variable-sensitivity/constant_array_abstract_object.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/constant_array_abstract_object.cpp]: Fixed.

File: src/analyses/variable-sensitivity/abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_object.h:47]: It is following the https://en.wikipedia.org/wiki/Prototype_pattern .  There is another good question over whether it is really needed.  Once the factory is refactored it should be clearer if and how it is used.

File: src/analyses/variable-sensitivity/data_dependency_context.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/data_dependency_context.cpp]: Fixed.

File: src/analyses/variable-sensitivity/full_struct_abstract_object.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/full_struct_abstract_object.cpp]: Done.

File: src/analyses/variable-sensitivity/interval_abstract_value.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/interval_abstract_value.cpp:214]: Fair point.  We do need to change how things are evaluated and the dispatch for that.  I hope that might simplify things a bit.

File: src/analyses/variable-sensitivity/interval_array_abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/interval_array_abstract_object.h:29]: Yeah, as above, I don't know if we can just remove it.

File: src/analyses/variable-sensitivity/struct_abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/struct_abstract_object.h:83]: Fair point but it is a pointer so it shouldn't cost much.

File: src/analyses/variable-sensitivity/struct_abstract_object.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/struct_abstract_object.h:133]: As above.

File: src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp:12]: I am not sure I follow.  This class is one of the few that actually is an abstract interpreter.  Almost all of the functionality is inherited from other classes but it really is different from the rest of the code in this directory as it isn't really part of VSD.

File: src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp]: Fixed.

File: src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp]: Ooops :-\  Removed!

File: src/analyses/variable-sensitivity/variable_sensitivity_domain.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/variable_sensitivity_domain.h]: Done.

File: src/analyses/variable-sensitivity/variable_sensitivity_object_factory.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/variable_sensitivity_object_factory.h:77]: The comment is by @thk123 so I fear the answer to that may have left with him.  I don't know and regard it as a "to fix".

File: src/analyses/variable-sensitivity/write_location_context.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/write_location_context.cpp]: Fixed.

File: src/analyses/variable-sensitivity/write_location_context.cpp
Comment (martin-cs) [reply src/analyses/variable-sensitivity/write_location_context.cpp]: Fixed.

File: unit/analyses/variable-sensitivity/abstract_object/merge.cpp
Comment (martin-cs) [reply unit/analyses/variable-sensitivity/abstract_object/merge.cpp]: Fixed.

File: unit/analyses/variable-sensitivity/constant_abstract_value/merge.cpp
Comment (martin-cs) [reply unit/analyses/variable-sensitivity/constant_abstract_value/merge.cpp]: Fixed.

File: unit/analyses/variable-sensitivity/constant_array_abstract_object/merge.cpp
Comment (martin-cs) [reply unit/analyses/variable-sensitivity/constant_array_abstract_object/merge.cpp]: Fixed.

File: unit/analyses/variable-sensitivity/eval.cpp
Comment (martin-cs) [reply unit/analyses/variable-sensitivity/eval.cpp]: Fixed.

**Comment by martin-cs** [comment]
@peterschrammel I need to move various doxygen comments into the headers but otherwise I should be done with the requested changes.  Let me know if there are more.

@chrisr-diffblue, @danpoe, @hannes-steffenhagen-diffblue : you know this code, any chance of a review?

File: src/analyses/variable-sensitivity/abstract_enviroment.h
Comment (martin-cs) [reply src/analyses/variable-sensitivity/abstract_enviroment.h:115]: Done.

**Comment by martin-cs** [comment]
@peterschrammel I have moved the doxygen comments to the headers.  Modulo fixing any issues doxygen has, I think I'm done.

**Comment by martin-cs** [comment]
@peterschrammel Are the changes OK?

 @danpoe @hannes-steffenhagen-diffblue @smowton you are owners for this area, please can you say something about this.  I don't think an in-depth review is appropriate but some kind of comment would help.

@owen-mc-diffblue If I understand correctly, you were an owner under a different name that hasn't been updated in CODEOWNERS.  Please can you comment?

**Comment by peterschrammel** [comment]
@martin-cs, lgtm modulo the doxygen issues: https://travis-ci.org/github/diffblue/cbmc/jobs/725557168

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
Thanks @peterschrammel I will work through the doxygen issues once I have it set up on my machine.

**Comment by martin-cs** [comment]
The doxygen issues should be finished.  I think it's just waiting on a stalled CI job.

**Comment by martin-cs** [comment]
@NlightNFotis I'm seeing build failures on MacOS.  But when I try to get the info about them I get:

"Error: We are currently unable to download the log. Please try again later. "

Any thoughts?  Did your PR fix everything?

**Comment by martin-cs** [comment]
@peterschrammel @tautschnig @chrisr-diffblue @smowton I can't seem to access the build logs for a failing github action

`pull-request-checks / check-macos-10_15-cmake-clang (pull_request)`

any chance of some help?

**Comment by peterschrammel** [comment]
@martin-cs, you should be able to see the log when you click on "Details" next to the failed job (provided that you are logged into github).

**Comment by martin-cs** [comment]
More specifically,

https://github.com/diffblue/cbmc/pull/5472/checks?check_run_id=1379416806

says

```
Error: We are currently unable to download the log. Please try again later.
```

I've tried the procedure given in https://github.com/diffblue/cbmc/blob/develop/AccessingCodebuildLogs.md but it doesn't seem to work.

**Comment by martin-cs** [comment]
Thanks @peterschrammel I am logged in and have tried that.  Any more suggestions?  Can you see it?

**Comment by peterschrammel** [comment]
I can see it. You have all permissions to access it. So, either a temporary Github issue or a problem with your browser.

**Comment by martin-cs** [comment]
Thanks @peterschrammel that's weird...  I will investigate.  What permissions do people need to see these?

**Comment by peterschrammel** [comment]
It seems for public repos the only requirement is that you are logged in. No org membership or repo permissions required.

**Comment by martin-cs** [comment]
Cool; have got it working (browser issue) thanks for the prompt support.

**Comment by NlightNFotis** [comment]
This PR looks to be in a good state, and it's mergeable. Is there anything outstanding on it? If not, I can merge it now.

**Comment by martin-cs** [comment]
@NlightNFotis  Thanks.  There are still quite a few refactorings, fixes, improvements, etc. that we will be making but TBH I would rather merge this and have these as separate PRs so they are easier to review.  The code in this PR works, has been deployed, has a higher test coverage than the rest of the project -- let's get it merged.  Plus it would be good to acknowledge everyone's hard work on this.

While we are on the subject, do you / @hannes-steffenhagen-diffblue / @chrisr-diffblue / @peterschrammel want / need any of the rest of the code from this branch?  I have a branch with it in, partially ready to be converted into PRs and I'm happy to work with people to do that but it is not a high priority if it is just me.

**Comment by peterschrammel** [comment]
@martin-cs, it's favourable to get any (valuable) code that is lying around in branches in various forks into the shape of PRs eventually. It's shame when a lot of work has been done and then it rots and dies... and someone else starts reimplementing something similar...

**Comment by martin-cs** [comment]
@peterschrammel agreed it is a question of priorities and time.  I could make them into draft PRs but they are unlikely to make mergable without someone else getting involved.

**Comment by NlightNFotis** [comment]
@martin-cs I think there is value for the draft PRs to be made - if we can find some time we can shape them up to be more mergeable.

Obviously, I cannot give guarantees for when that is, as it depends on a lot of things, but if the PRs are there, they are much more likely to be picked up by someone.

**Comment by thk123** [comment]
😮  Never thought I'd see the day - can't let it go without commenting! 🎉 🎉 🎉

Hope you're all well

**Comment by smowton** [comment]
Signs of the end times:
[x] POTUS refusing to leave the white house
[x] The variable-sensitivity domain has been merged
[ ] Rains of branston pickle

**Comment by chrisr-diffblue** [comment]
This really should have been merged on 1st April....

**Comment by hannes-steffenhagen-diffblue** [comment]
Hello @thk123 and @smowton, nice to hear from you! I hope you're all well too!

**Comment by martin-cs** [comment]
@chrisr-diffblue ... I should have thought of that ... *sigh*

**Comment by peterschrammel** [comment]
11/11 11:11 is the beginning of carnival...

**Pull request merged**
